sway/desktop/output.c

```diff
@@ -224,11 +224,146 @@ static void render_view(struct sway_view *view, struct sway_output *output) {
 }
 
 /**
- * Render decorations for a view with "border normal".
+ * Select a color to use to render a border based on focus.
+ */
+static struct border_colors *select_focus_color(char *name) {
+	struct border_colors *colors;
+	if (strcmp(name, "focused") == 0) {
+		return colors = &config->border_colors.focused;
+	} else if (strcmp(name, "inactive") == 0) {
+		return colors = &config->border_colors.focused_inactive;
+	} else if (strcmp(name, "unfocused") == 0) {
+		return colors =&config->border_colors.unfocused;
+	} else {
+		return NULL;
+	}
+}
+
+/**
+ * Get all title textures from children of a parent.
+ */
+static struct wlr_texture **get_title_textures(struct sway_container *con) { 
+	struct sway_seat *seat = input_manager_current_seat(input_manager);
+	struct sway_container *focus = seat_get_focus(seat);
+	int num_tabs = con->children->length;
+	struct wlr_texture **title_texture = 
+		malloc(num_tabs * sizeof(struct wlr_texture));
+
+	for (int i = 0; i < con->children->length; ++i) {
+		struct sway_container *child = con->children->items[i];
+
+		if (child->type == C_VIEW) {
+			if (child->sway_view->border != B_NONE) {
+				if (focus == child) {
+					title_texture[i] = child->title_focused;
+				} else if (seat_get_focus_inactive(seat, con) == child) {
+					title_texture[i] = child->title_focused_inactive;
+				} else {
+					title_texture[i] = child->title_unfocused;
+				}
+			}
+		}
+	}
+	return title_texture;
+}
+
+/**
+ * Render decorations for the top border with "border normal".
+ */
+static void render_container_top_border_normal(struct sway_output *output,
+		struct sway_container *con, struct border_colors *colors,
+		struct wlr_texture *title_texture, int depth) {
+
+	struct wlr_renderer *renderer =
+		wlr_backend_get_renderer(output->wlr_output->backend);
+	struct wlr_box box;
+	float color[4];
+	size_t num_tabs = 1;
+	size_t tab_width = 1;
+	
+	if (con->parent->layout == L_TABBED) {
+		num_tabs = con->parent->children->length;
+		tab_width = (con->width / num_tabs);
+	}
+
+	// Single pixel bar above title
+	memcpy(&color, colors->border, sizeof(float) * 4);
+	color[3] *= con->alpha;
+	box.x = con->x + depth * tab_width;
+	box.y = con->y;
+	box.width = con->width / num_tabs;
+	box.height = 1;
+	scale_box(&box, output->wlr_output->scale);
+	wlr_render_rect(renderer, &box, color,
+			output->wlr_output->transform_matrix);
+
+	// Single pixel bar below title
+	box.x = con->x + depth * tab_width;
+	box.y = con->sway_view->y - 1;
+	box.width = con->width / num_tabs;
+	box.height = 1;
+	scale_box(&box, output->wlr_output->scale);
+	wlr_render_rect(renderer, &box, color,
+			output->wlr_output->transform_matrix);
+
+	// Title background
+	memcpy(&color, colors->background, sizeof(float) * 4);
+	color[3] *= con->alpha;
+	box.x = con->x + depth * tab_width;
+	box.y = con->y + 1;
+	box.width = con->width / num_tabs + 2;
```

Comment:
I don't think this is right. The width needs to account for the borders on the sides. The title texture overlaps the right border because of this.